crates/containerd-shim-wasmtime/src/instance.rs

```diff
@@ -81,200 +46,178 @@ fn load_spec(bundle: String) -> Result<oci::Spec, Error> {
 
-        debug!("getting start function");
-        let f = i.get_func(&mut store, &m.2).ok_or_else(|| {
-            Error::InvalidArgument("module does not have a wasi start function".to_string())
-        })?;
+        let mut container = self.build_container(
+            self.stdin.as_str(),
+            self.stdout.as_str(),
+            self.stderr.as_str(),
+            engine,
+        )?;
 
-        debug!("starting wasi instance");
-
-        let cg = oci::get_cgroup(&spec)?;
-
-        oci::setup_cgroup(cg.as_ref(), &spec)
-            .map_err(|e| Error::Others(format!("error setting up cgroups: {}", e)))?;
-
-        let res = unsafe { exec::fork(Some(cg.as_ref())) }?;
-        match res {
-            exec::Context::Parent(tid, pidfd) => {
-                let mut lr = self.pidfd.lock().unwrap();
-                *lr = Some(pidfd.clone());
-
-                debug!("started wasi instance with tid {}", tid);
-
-                let code = self.exit_code.clone();
-
-                let _ = thread::spawn(move || {
-                    let (lock, cvar) = &*code;
-                    let status = match pidfd.wait() {
-                        Ok(status) => status,
-                        Err(e) => {
-                            error!("error waiting for pid {}: {}", tid, e);
-                            cvar.notify_all();
-                            return;
-                        }
-                    };
-
-                    debug!("wasi instance exited with status {}", status.status);
-                    let mut ec = lock.lock().unwrap();
-                    *ec = Some((status.status, Utc::now()));
-                    drop(ec);
-                    cvar.notify_all();
-                });
-                Ok(tid)
-            }
-            exec::Context::Child => {
-                // child process
-
-                // TODO: How to get exit code?
-                // This was relatively straight forward in go, but wasi and wasmtime are totally separate things in rust.
-                match f.call(&mut store, &[], &mut []) {
-                    Ok(_) => std::process::exit(0),
-                    Err(_) => std::process::exit(137),
-                };
-            }
-        }
+        log::info!("created container: {}", self.id);
+        let code = self.exit_code.clone();
+        let pid = container.pid().unwrap();
+
+        container
+            .start()
+            .map_err(|err| Error::Any(anyhow::anyhow!("failed to start container: {}", err)))?;
+
+        thread::spawn(move || {
+            let (lock, cvar) = &*code;
+
+            let status = match waitid(WaitID::Pid(pid), WaitPidFlag::WEXITED) {
+                Ok(WaitStatus::Exited(_, status)) => status,
+                Ok(WaitStatus::Signaled(_, sig, _)) => sig as i32,
+                Ok(_) => 0,
+                Err(e) => {
+                    if e == Errno::ECHILD {
+                        0
```

Comment:
I assume this code came from my code, but it has to plant the log in this case.

Reply:
Could you elaborate more about "plant the log"? Do you want me to log a message when `Errno::ECHILD` was returned?

Reply:
>  when Errno::ECHILD was returned?

Yes 👍

Reply:
I have added a log messaging saying "no child process" when `Errno::ECHILD` was returned in both the shims.

Reply:
I am not entirely sure when people see this log message, that actions they can do to mitigate. Or is this an issue at all in the first place? I would appreciate if anyone can help me to understand it.